README

Bristol Emulations
------------------

This is a write-up of each of the emulated synthesisers. The algorithms
employed were 'gleaned' from a variety of sources including the original
owners manuals, so they may be a better source of information. The author
has owned and used a selection but far from all of the originals. Some of them
were built just from descriptions of their operation, or from understanding
how synths work - most of them were based on the Mini Moog anyway. Many of
the synths share components: the filter covers most of them, the Prophets and
Oberheims share a common oscillator and the same LFO is used in many of them.
Having said that each one differs considerably in the resulting sound that is
generated, more so than initially expected. Each release refines each of the
components and the result is that all emulations benefit from the improvements.
All the emulations have distinctive sounds, not least due to that the original
instruments used different modulations and mod routing.
The filter, which is a large defining factor in the tonal qualities of any
synth, is common to all the emulations. The filter implements a few different 
algorithms and these do separate each of the synths: the Explorer layering
two low pass filters on top of each other: the OB-Xa using different types
depending on 'Pole' selection. Since release 0.20.8 the emulator has had a
Houvillainen non-linear ladder filter integrated which massively improves 
the quality at considerable expense to the CPU.
There is one further filter algorithm used solely for the Leslie rotary 
emulator crossover, this is a butterworth type filter.

Bristol is in no way related to any of the original manufacturers whose 
products are emulated by the engine and represented by the user interface,
bristol does not suggest that the emulation is a like representation of the
original instrument, and the author maintains that if you want the original
sound then you are advised to seek out the original product. Alternatively a
number of the original manufacturers now provide their own vintage collections
which are anticipated to be more authentic. All names and trademarks used by
Bristol are ownership of the respective companies and it is not inteded to 
misappropriate their use here. If you have concerns you are kindly requested
to contact the author.

The write-up includes the parameter operations, modulations, a description of
the original instrument and a brief list of the kind of sounds you can expect
by describing a few of the well known users of the synth.

Several emulations have not been written up. Since the APR 2600 was implemented
it became a rather large job to actually describe what was going on. If you 
really want to know about the synths that are not in this document then you
might want to search for their owners manuals.

All emulations are available from the same engine, just launch multiple GUIs
and adjust the midi channels for multi timbrality and layering.

It is noted here that the engine is relatively 'dumb'. Ok, it generates a very
broad range of sounds, currently about 25 different synthesisers and organs,
but it is not really intelligent. Memories are a part of the GUI specification
- it tells the engine which algorithm to use on which MIDI channel, then it
calls a memory routine that configures all the GUI controllers and a side effect
of setting the controllers is that their values are sent to the engine. This is
arguably the correct model but it can affect the use of MIDI master keyboards.
The reason is that the GUI is really just a master keyboard for the engine and
drives it with MIDI SYSEX messages over TCP sessions. If you were to alter the
keyboard transpose, for example, this would result in the GUI sending different
'key' numbers to the engine when you press a note. If you were already driving
the synth from a master keyboard then the transpose button in the Brighton GUI
would have no effect - the master keyboard would have to be transposed instead.
This apparent anomaly is exacerbated by the fact that some parameters still are
in the engine, for example master tuning is in the engine for the pure fact that
MIDI does not have a very good concept of master tuning (only autotuning).
Irrespective of this, bristol is a synthesiser so it needs to be played, 
tweaked, driven. If you think that any of the behaviour is anomalous then let
me know. One known issue in this area is that if you press a key, transpose
the GUI, then release the key - it will not go off in the engine since the GUI
sends a different key code for the note off event - the transposed key. This
cannot be related to the original keypress. This could be fixed with a MIDI all
notes off event on 'transpose', but I don't like them. Also, since the 0.20
stream the problem only affects a few of the emulations, the rest now sending
a transpose message to the engine and letting it do the work.

Since release 0.30.6 the engine correctly implements monophonic note logic.
Prior to this the whole engine was polyphonic and playing with one voice only
gave last note preference which dramatically affects playing styles - none of
the cool legato effects of the early monophonics. The quoted release fix this
limitation where the engine will keep a keymap of all played keys (one per
emulation) when started with a single voice and uses this map to provide
consistent note precedence, high note logic, low note logic or just using the
previously implemented last note logic. In this release the keymap was only
maintained with monophonic emulations, this is a potential extension as even
in polyphonic mode it would be useful for arpeggiation (which is currently
implemented using a FIFO rather than an ordered keymap).




    Moog Mini
    ---------

It is perhaps not possible to write up who used this synth, the list is endless.
Popular as it was about the first non-modular synthesiser, built as a fixed
configuration of the racked or modular predecessors.

Best known at the time on Pink Floyd 'Dark Side of the Moon' and other albums.
Rick Wakeman used it as did Jean Michel Jarre. Wakefield could actually
predict the sound it would make by just looking at the settings, nice to be
able to do if a little unproductive but it went to show how this was treated
as an instrument in its own right. It takes a bit of work to get the same sweet,
rich sounds out of the emulation, but it can be done with suitable tweaking.

The original was monophonic, although a polyphonic version was eventually made
after Moog sold the company - the MultiMoog. This emulation is more comparable
to that model as the sound is a bit thinner and can be polyphonic. The design
of this synth became the pole bearer for the following generations: it had 
three oscillators, one of which could become a low frequency modulator. They
were fed into a mixer with a noise source, and were then fed into a filter
with 2 envelope generators to contour the wave. Modulation capabilities were
not extensive, but interestingly enough it did have a frequency modulation (FM)
capability, eventually used by Yamaha to revolutionise the synthesiser market
starting the downfall of analogue synthesis twenty years later.

All the analogue synths were temperature sensitive. It was not unusual for the
synths to 'detune' between sound test and performance as the evening set in.
To overcome this they could optionally produce a stable A-440Hz signal for 
tuning the oscillators manually - eventually being an automated option in the
newer synths. Whilst this digital version has stable frequency generation the
A-440 is still employed here for the sake of it.

Modifiers and mod routing are relatively simple, Osc-3 and noise can be mixed,
and this signal routed to the oscillator 1 and 2 frequency or filter cutoff.

The synth had 5 main stages as follows:

Control:

    Master tuning: up/down one note.

    Glide: (glissando, portamento). The rate at which one key will change its
    frequency to the next played key, 0 to 30 seconds.

    Mod: source changes between Osc-3 and noise.

    Release: The envelope generators had only 3 parameters. This governed whether
    a key would release immediately or would use Decay to die out.

    Multi: Controls whether the envelope will retrigger for each new keypress.

Oscillators:

    There are three oscillators. One and two are keyboard tracking, the third
    can be decoupled and used as an LFO modulation source.

    Oscillator 1:
        Octave step from 32' to 1'.
        Waveform selection: sine/square/pulse/ramp/tri/splitramp
        Mod: controls whether Osc-3/noise modulates frequency

    Oscillator 2:
        Octave step from 32' to 1'.
        Fine tune up/down 7 half notes.
        Waveform selection: sine/square/pulse/ramp/tri/splitramp
        Mod: controls whether Osc-3/noise modulates frequency

    Oscillator 3:
        Octave step from 32' to 1'.
        Fine tune up/down 7 half notes.
        Waveform selection: sine/square/pulse/ramp/tri/splitramp
        LFO switch to decouple from keytracking.

Mixer:

    Gain levels for Oscillator 1/2/3
    Mixing of the external input source into filter
    Noise source with white/pink switch.

    Note: The level at which Osc-3 and noise modulates sound depends on its
    gain here, similarly the noise. The modulator mix also affects this, but
    allows Osc-3 to mod as well as sound. The modwheel also affect depth.

Filter:

    Cutoff frequency

    Emphasis (affects Q and resonance of filter).

    Contour: defines how much the filter envelope affects cutoff.

    Mod - Keyboard tracking of cutoff frequency.

    Mod - Osc-3/noise modulation of cutoff frequency.

Contour:

    The synth had two envelope generators, one for the filter and one for the
    amplifier. Release is affected by the release switch. If off the the sound
    will release at the rate of the decay control.

    Attack: initial ramp up of gain.

    Decay: fall off of maximum signal down to:

    Sustain: gain level for constant key-on level.

    Key: Touch sensitivity of amplifier envelope.

Improvements to the Mini would be some better oscillator waveforms, plus an
alternative filter as this is a relatively simple synthesiser and could do
with a warmer filter (this was fixed with integration of the houvillanen filters
although the do consume a lot of CPU to do it).

The Output selection has a Midi channel up/down selector and memory selector.
To read a memory either use the up/down arrows to go to the next available
memory, or type in a 3 digit number on the telephone keypad and press 'L' for
load or 'S' for save.

As of release 0.20.5 Vasiliy Basic contributed his Mini memory banks and they
are now a part of the distribution:

Programs for Bristol's "Mini" (from 50 to 86 PRG)

List of programs:

    -Melodic-
    50 - Trumpet
    51 - Cello
    52 - Guitar 1
    53 - Guitar 2
    54 - Fingered Bass
    55 - Picked Bass
    56 - Harmonica
    57 - Accordion
    58 - Tango Accordion
    59 - Super Accordion
    60 - Piano
    61 - Dark Organ
    62 - Flute
    63 - Music Box
    64 - Glass Xylo
    65 - Glass Pad
    66 - Acid Bass
    
    -Drums-
    67 - Bass Drum 1 
    68 - Bass Drum 2
    69 - Bass Drum 3
    70 - Bass Drum 4
    71 - Tom
    72 - Snare 1
    73 - Snare 2
    74 - Snare 3
    75 - Snare 4
    76 - Cl HH 1
    77 - Op HH 1
    78 - Crash Cym 1
    79 - Crash Cym 2
    80 - Cl HH 2
    81 - Op HH 2
    
    -FX-
    82 - Sea Shore
    83 - Helicopter 1
    84 - Helicopter 2
    85 - Bird Tweet
    86 - Birds Tweet




    Sequential Circuits Prophet-5
    Sequential Circuits Prophet-52 (the '5' with chorus)
    ----------------------------------------------------

Sequential circuits released amongst the first truly polyphonic synthesisers
where a group of voice circuits (5 in this case) were linked to an onboard
computer that gave the same parameters to each voice and drove the notes to

NEWS

    0.60.11 24 Jan 2013 Maintenance release
    0.60.10 27 Apr 2012 Maintenance release
    0.60.9 25 Oct 2011 Maintenance release, fanning linear potentiometers
    0.60.8 23 Dec 2010 Maintenance release, Hammond optimisations.
    0.60.7 22 Oct 2010 Maintenance release, minor features
    0.60.6 11 Aug 2010 Geometry configuration options, maintenance release
    0.60.5 08 Jun 2010 Maintenance release
    0.60.4 23 May 2010 Maintenance release
    0.60.3 12 May 2010 Maintenance release
    0.60.2 28 Apr 2010 Command Line Interface Changes, maintenance release
    0.60.1 23 Apr 2010 Command Line Interface Changes
    0.60.0 19 Apr 2010 Command Line Interface

    0.50.8 05 May 2010 Maintenance updates
    0.50.7 22 Apr 2010 CV Keyboard Tracking Frequency Control for ARP
    0.50.6 08 Apr 2010 Multiple IO Channels for audio/CV, more NRP Controller Support, Jack Session Manager maintenance
    0.50.5 03 Apr 2010 Jack Session Manager Immediate fixes
    0.50.4 02 Apr 2010 Jack Session Manager Support
    0.50.3 17 Mar 2010 Band limited oscillator corrections, filter optimisation, GUI enhancements.
    0.50.2 06 Feb 2010 Real time processing enhancements
    0.50.1 22 Jan 2010 Usability enhancments, LADI level 1 compliance, process distribution

    0.40.8 22 Jan 2010 Maintenance release
    0.40.7 22 Nov 2009 Maintenance release
    0.40.6 28 Sep 2009 Maintenance release
    0.40.5 23 Jul 2009 Distributed processing maintenance release
    0.40.4 06 Jun 2009 Audio Driver Mainenance Release
    0.40.3 25 May 2009 Maintenance release, B3 stuck notes.
    0.40.2 05 May 2009 Maintenance release, usability improvements.
    0.40.1 25 Apr 2009 Commodore C64 SID emulator

    0.30.9 02 Apr 2009 Bristol BassMaker, massive filter optimisations.
    0.30.8 20 Mar 2009 Maintenance release, ARP 2600 fixes.
    0.30.7 02 Mar 2009 Baumann BME-700
    0.30.6 22 Feb 2009 Monophonic note precedence key logic.
    0.30.5 01 Feb 2009 Maintenance release?
    0.30.4 20 Jan 2009 Korg Poly-800
    0.30.3 19 Dec 2008 Crumar Trident, PWM and Sync fixes, Polysix improvements
    0.30.2 15 Dec 2008 Crumar Stratus
    0.30.1 03 Dec 2008 Voyager Electric Blue, Moog Sonic 6.

    0.20.10 27 Nov 2008 Maintenance release.
    0.20.9 14 Nov 2008 Maintenance release.
    0.20.8 28 Sep 2008 New filters, normalised gains, diverse fixes
    0.20.7 29 Aug 2008 Sequential Circuits Pro-One, Scala tonal mapping tables, overhaul of the Prophet-10
    0.20.6 13 Jun 2008 Bandwidth limited oscllators. UI Heartbeats, GUI scaling.
    0.20.5 20 Apr 2008 First release of the Jupiter-8 emulation, GUI prealiasing, Vasiliy Basic's Mini memory suite.
    0.20.4 26 Mar 2008 Arpeggiating OBXa, fixes, synchronised threads.
    0.20.3 05 Mar 2008 Integrated a new Crumar Bit-100 emulation
    0.20.2 25 Feb 2008 Integrated a few Crumar Bit emulations.
    0.20.1 20 Jan 2008 First release of the 0.20 stream

    0.10.13 19 Jan 2008 Maintenance release.
    0.10.12 03 Dec 2007 Maintenance release.

    0.10.11 27 Oct 2007 Arpeggiator written into the MIDI library.

Added arpeggiation (actually an arpeggiating sequencer) into MIDI library.

    0.10.10 XX Oct 2007 Bass pedalboard integration for Hammond.

Added a third manual to the hammond.
Built in the GM-2 MIDI controller mappings to a range of the emulations.

    0.10.9 16 Sept 2007 Reworked installation procedures, Vox-300

Vox dual manual 300.

    0.10.8 01 Sept 2007 Realistic Concertmate, recolorations, fixes

Developed a Realistic/Moog Concertmate MG1.
Diverse fixes and recolorations.

    0.10.7 29 July 2007 added roadrunner electric piano

Two rounds of GUI optimisation have improved the CPU load by a couple of orders
of magnitude.

Added a grey scale option.

Added antialiasing features (that had been previously dropped as too CPU
intensive, now quite reasonable).

    0.10.5 06 June 2007 added roadrunner electric piano

The Roadrunner is an emulated organ circuit electric piano, many of these were
released in the 70's and complemented the string machines. They used clock 
divided organ circuitry and simple envelope generators per key to groom to 
something approaching a piano envelope.

Diverse minor bugfixes.

    0.10.4 24 May 2007 added Solina string machine

Potentials:
Some of the diverse synth memories are damaged.
ARP 2600 has damaged vertical patch cables.
Move the key mappings and controller mappings to .mcm file as they are static.
Removed clipping in new keyclick generation.

    0.10.3 18 May 2007 added the following

Modifications to configure.ac and Makefile.am to correct autotools errors.
Corrected use of --exec_prefix for configure process.

Note velocity, Poly and Channel Pressure curve mappings from file. Global.
Note integer index mapping in same configuration file. Global.
Note Microtonal tunings. Global settings - affects the whole engine.

Note velocity map - per emulation.
Note microtonal map - per emulation.

B3 key noise generated per bus with configurable delays and gain per contact.
Reverse controller motion of hammond drawbars (mappings?).

Controller mapped value curves for lin/log/exp/parabolic with inverse table.

Resolved issues with use of '-jack' flag as opposed to '-audio jack'.

More keyboard velocity curves of which one mapping per emulation.

Pitch bend damaged by microtonal mapping implementation, fixed.

    0.10.2: 10 May 2007

Changes to automake configuration files for installation prefixes.

    0.10.1: 4 May 2007

Probable first distribution of the bristol version using GNU autotools.



bristol/dco.c

/*
 *  Diverse Bristol audio routines.
 *  Copyright (c) by Nick Copeland <nickycopeland@hotmail.com> 1996,2012
 *
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, see <http://www.gnu.org/licenses/>.
 *
 */
/*#define BRISTOL_DBG */
/*
 * Need to have basic template for an operator. Will consist of
 *
 *	dcoinit()
 *	operate()
 *	reset()
 *	destroy()
 *
 *	destroy() is in the library.
 *
 * Operate will be called when all the inputs have been loaded, and the result
 * will be an output buffer written to the next operator.
 */

#include <math.h>

#include "bristol.h"
#include "bristolblo.h"
#include "dco.h"

float note_diff;

/*
 * The name of this operator, IO count, and IO names.
 */
#define OPNAME "DCO"
#define OPDESCRIPTION "Digitally Controlled Oscillator"
#define PCOUNT 2
#define IOCOUNT 2
/* static char *IO[] = {"FreqMod", "Output"}; */

#define DCO_IN_IND 0
#define DCO_OUT_IND 1

#define DCO_WAVE_COUNT 6

static void fillWave();

/*
 * Reset any local memory information.
 */
static int destroy(bristolOP *operator)
{
#ifdef BRISTOL_DBG
	printf("destroy(%x)\n", operator);
#endif

	/*
	 * Unmalloc anything we added to this structure
	 */
	bristolfree(((bristolDCO *) operator)->wave[0]);
	bristolfree(((bristolDCO *) operator)->wave[1]);
	bristolfree(((bristolDCO *) operator)->wave[2]);
	bristolfree(((bristolDCO *) operator)->wave[3]);
	bristolfree(((bristolDCO *) operator)->wave[4]);
	bristolfree(((bristolDCO *) operator)->wave[5]);
	bristolfree(((bristolDCO *) operator)->wave[6]);
	bristolfree(((bristolDCO *) operator)->wave[7]);

	bristolfree(operator->specs);

	/*
	 * Free any local memory. We should also free ourselves, since we did the
	 * initial allocation.
	 */
	cleanup(operator);
	return(0);
}

/*
 * Reset any local memory information.
 */
static int reset(bristolOP *operator, bristolOPParams *param)
{
#ifdef BRISTOL_DBG
	printf("dcoreset(%x)\n", operator);
#endif

	if (param->param[0].mem != NULL)
		bristolfree(param->param[0].mem);
	param->param[0].mem = bristolmalloc0(sizeof(float) * DCO_WAVE_SZE);
	param->param[0].int_val = 0;
	param->param[1].float_val = 0.5;
	param->param[2].float_val = 1.0;
	param->param[3].float_val = 1.0;
	return(0);
}

/*
 * Alter an internal parameter of an operator.
 */
static int param(bristolOP *operator, bristolOPParams *param,
	unsigned char index, float value)
{
#ifdef BRISTOL_DBG
	printf("dcoparam(%x, %x, %i, %f)\n", operator, param, index, value);
#endif

	switch (index) {
		case 0:
			if ((param->param[index].int_val = value * CONTROLLER_RANGE)
				> DCO_WAVE_COUNT)
				param->param[index].int_val = DCO_WAVE_COUNT;
			break;
		case 1:
			/*
			 * Transpose in octaves. 0 to 5, down 2 up 3.
			 */
			{
				int note = value * CONTROLLER_RANGE;

				switch (note) {
					case 0:
						param->param[1].float_val = 0.25;
						break;
					case 1:
						param->param[1].float_val = 0.50;
						break;
					case 2:
						param->param[1].float_val = 1.0;
						break;
					case 3:
						param->param[1].float_val = 2.0;
						break;
					case 4:
						param->param[1].float_val = 4.0;
						break;
					case 5:
						param->param[1].float_val = 8.0;
						break;
				}
			}
			break;
		case 2: /* Tune in "small" amounts */
			{
				float tune, notes = 1.0;
				int i;

				tune = (value - 0.5) * 2;

				/*
				 * Up or down 7 notes.
				 */
				for (i = 0; i < 7;i++)
				{
					if (tune > 0)
						notes *= note_diff;
					else
						notes /= note_diff;
				}

				if (tune > 0)
					param->param[index].float_val =
						1.0 + (notes - 1) * tune;
				else
					param->param[index].float_val =
						1.0 - (1 - notes) * -tune;

				break;
			}
		case 3:
			param->param[index].float_val = value;
			break;
	}

	return(0);
}

/*
 * As of the first write, 9/11/01, this takes nearly 20mips to operate a single
 * oscillator. Will work on optimisation of the code, using non-referenced 
 * variables in register workspace.
 *
 *	output buffer pointer.
 *	output buffer index.
 *	input buffer index.
 *	wavetable.
 *	wavetable index.
 *	count.
 *	gain.
 *
 * With optimisations this is reduced to a nominal amount. Put most parameters
 * in registers, and then stretched the for loop to multiples of 16 samples.
 *
 * I would prefer to have this turned into a linear input, rated at 1float per
 * octave. This would require some considerable alterations to the oscillator
 * sample step code, though. The basic operation is reasonably simple - we take
 * a (midikey/12) and it gives us a step rate. We need to work in portamento and
 * some more subtle oscillator driver changes.
 */
static int operate(bristolOP *operator,
	bristolVoice *voice,
	bristolOPParams *param,
	void *lcl)
{
	bristolDCOlocal *local = lcl;
	register int obp, count;
	register float *ib, *ob, *wt, wtp, gain, transp;
	bristolDCO *specs;

	specs = (bristolDCO *) operator->specs;

#ifdef BRISTOL_DBG
	printf("dco(%x, %x, %x)\n", operator, param, local);
#endif

	count = specs->spec.io[DCO_OUT_IND].samplecount;
	ib = specs->spec.io[DCO_IN_IND].buf;
	ob = specs->spec.io[DCO_OUT_IND].buf;
	wtp = local->wtp;
	gain = param->param[3].float_val;
	transp = param->param[1].float_val * param->param[2].float_val;
	wtp = local->wtp;

 	if (bristolBLOcheck(voice->cFreq*transp)) {
		wt = param->param[0].mem;
		memset(wt, 0, DCO_WAVE_SZE * sizeof(float));
		switch (param->param[0].int_val) {
			case 0:
				/* Sine, cannot BWL this */
				wt = specs->wave[param->param[0].int_val];
				break;
			case 1:
				generateBLOwaveformF(voice->cFreq*transp, wt, BLO_SQUARE);
				break;
			case 2:
				/* This will be pulse */
				generateBLOwaveformF(voice->cFreq*transp, wt, BLO_PULSE);
				break;
			case 3:
				generateBLOwaveformF(voice->cFreq*transp, wt, BLO_SAW);
				break;
			case 4:
				generateBLOwaveformF(voice->cFreq*transp, wt, BLO_TRI);
				break;
			case 5:
			{
				float ws[1024];
				memset(ws, 0, DCO_WAVE_SZE * sizeof(float));
				generateBLOwaveformF(voice->cFreq*transp, &ws[0], BLO_SAW);

				for (obp = 0; obp < 1024; obp++)
					wt[obp] = (ws[obp] + ws[(obp * 2) & 1023]) * 0.5;

				break;
			}
			case 6:
				generateBLOwaveformF(voice->cFreq*transp, wt, BLO_RAMP);
				break;
			case 7:
				generateBLOwaveformF(voice->cFreq*transp, wt, BLO_RAMP);
				break;
		}
	} else
		wt = specs->wave[param->param[0].int_val];

/*printf("%i, %f, %i: %x %x %x\n", count, gain, param->param[0].int_val, wt, ib, ob); */
	/*
	 * Go jumping through the wavetable, with each jump defined by the value
	 * given on our input line, making sure we fill one output buffer.
	 */
	for (obp = 0; obp < count; obp++)
	{
		if (wtp >= DCO_WAVE_SZE_M)
			ob[obp] += (wt[0] * (wtp - ((float) ((int) wtp)))
				+ wt[(int) wtp] * (1.0 - (wtp - ((float) ((int) wtp))))) * gain;
		else
			ob[obp] += (wt[(int) wtp + 1] * (wtp - ((float) ((int) wtp)))
				+ wt[(int) wtp] * (1.0 - (wtp - ((float) ((int) wtp))))) * gain;

		/*
		 * Move the wavetable pointer forward by amount indicated in input 
		 * buffer for this sample.
		 */
		if ((wtp += ib[obp] * transp) >= DCO_WAVE_SZE)
		{
			while (wtp >= DCO_WAVE_SZE)
				wtp -= DCO_WAVE_SZE;
		}
		/*
		 * If we have gone negative, round back up. Allows us to run the 
		 * oscillator backwards.
		 */
		while (wtp < 0)
			wtp += DCO_WAVE_SZE;
	}

	local->wtp = wtp;
	return(0);
}

/*
 * Setup any variables in our OP structure, in our IO structures, and malloc
 * any memory we need.
 */
bristolOP *
dcoinit(bristolOP **operator, int index, int samplerate, int samplecount)
{
	bristolDCO *specs;

	*operator = bristolOPinit(operator, index, samplecount);

#ifdef BRISTOL_DBG
	printf("dcoinit(%x(%x), %i, %i, %i)\n",
		operator, *operator, index, samplerate, samplecount);
#endif

	note_diff = pow(2, ((double) 1)/12);

	/*
	 * Then the local parameters specific to this operator. These will be
	 * the same for each operator, but must be inited in the local code.
	 */
	(*operator)->operate = operate;
	(*operator)->destroy = destroy;
	(*operator)->reset = reset;
	(*operator)->param= param;

	specs = (bristolDCO *) bristolmalloc0(sizeof(bristolDCO));
	(*operator)->specs = (bristolOPSpec *) specs;
	(*operator)->size = sizeof(bristolDCO);

	/*
	 * These are specific to this operator, and will need to be altered for
	 * each operator.
	 */
	specs->spec.opname = OPNAME;
	specs->spec.description = OPDESCRIPTION;
	specs->spec.pcount = PCOUNT;
	specs->spec.iocount = IOCOUNT;
	specs->spec.localsize = sizeof(bristolDCOlocal);

	/*
	 * We are going to assign multiple waves to this oscillator.
	 * sine, ramp, square, triangle?
	 */
	specs->wave[0] = (float *) bristolmalloc(DCO_WAVE_SZE * sizeof(float));
	specs->wave[1] = (float *) bristolmalloc(DCO_WAVE_SZE * sizeof(float));
	specs->wave[2] = (float *) bristolmalloc(DCO_WAVE_SZE * sizeof(float));
	specs->wave[3] = (float *) bristolmalloc(DCO_WAVE_SZE * sizeof(float));
	specs->wave[4] = (float *) bristolmalloc(DCO_WAVE_SZE * sizeof(float));
	specs->wave[5] = (float *) bristolmalloc(DCO_WAVE_SZE * sizeof(float));
	specs->wave[6] = (float *) bristolmalloc(DCO_WAVE_SZE * sizeof(float));
	specs->wave[7] = (float *) bristolmalloc(DCO_WAVE_SZE * sizeof(float));

	/*
	 * FillWave is something that should be called as a parameter change, but
	 * for testing will load it here.
	 */
	fillWave(specs->wave[0], DCO_WAVE_SZE, 0);
	fillWave(specs->wave[1], DCO_WAVE_SZE, 1);
	fillWave(specs->wave[2], DCO_WAVE_SZE, 2);
	fillWave(specs->wave[3], DCO_WAVE_SZE, 3);
	fillWave(specs->wave[4], DCO_WAVE_SZE, 4);
	fillWave(specs->wave[5], DCO_WAVE_SZE, 5);
	fillWave(specs->wave[6], DCO_WAVE_SZE, 6);
	fillWave(specs->wave[7], DCO_WAVE_SZE, 7);

	/*
	 * Now fill in the dco specs for this operator. These are specific to an
	 * oscillator.
	 */
	specs->spec.param[0].pname = "waveform";
	specs->spec.param[0].description= "sine square pulse ramp triangle tan saw";
	specs->spec.param[0].type = BRISTOL_ENUM;
	specs->spec.param[0].low = 0;
	specs->spec.param[0].high = 5;
	specs->spec.param[0].flags = BRISTOL_ROTARY|BRISTOL_SLIDER;

	specs->spec.param[1].pname = "Transpose";
	specs->spec.param[1].description = "Octave transposition";
	specs->spec.param[1].type = BRISTOL_FLOAT;
	specs->spec.param[1].low = 0;
	specs->spec.param[1].high = 12;
	specs->spec.param[1].flags = BRISTOL_ROTARY|BRISTOL_SLIDER;

	specs->spec.param[2].pname = "Tune";
	specs->spec.param[2].description = "fine tuning of frequency";
	specs->spec.param[2].type = BRISTOL_INT;
	specs->spec.param[2].low = 0;
	specs->spec.param[2].high = 127;
	specs->spec.param[2].flags = BRISTOL_ROTARY|BRISTOL_SLIDER;

	specs->spec.param[3].pname = "gain";
	specs->spec.param[3].description = "output gain on signal";
	specs->spec.param[3].type = BRISTOL_FLOAT;
	specs->spec.param[3].low = 0;
	specs->spec.param[3].high = 2;
	specs->spec.param[3].flags = BRISTOL_ROTARY|BRISTOL_SLIDER|BRISTOL_HIDE;

	/*
	 * Now fill in the dco IO specs.
	 */
	specs->spec.io[0].ioname = "input";
	specs->spec.io[0].description = "input rate modulation signal";
	specs->spec.io[0].samplerate = samplerate;
	specs->spec.io[0].samplecount = samplecount;
	specs->spec.io[0].flags = BRISTOL_DC|BRISTOL_INPUT;

	specs->spec.io[1].ioname = "output";
	specs->spec.io[1].description = "oscillator output signal";
	specs->spec.io[1].samplerate = samplerate;
	specs->spec.io[1].samplecount = samplecount;
	specs->spec.io[1].flags = BRISTOL_AC|BRISTOL_OUTPUT;

	return(*operator);
}

/*
 * Waves have a range of 24, which is basically two octaves. For larger 
 * differences will have to apply apms. The diverse leading edge wave have a
 * rather raw feeling and should be smoothed out a bit.
 */
static void
fillWave(float *mem, int count, int type)
{
	int i;

#ifdef BRISTOL_DBG
	printf("fillWave(%x, %i, %i)\n", mem, count, type);
#endif

	switch (type) {
		case 0:
			/*
			 * This will be a sine wave. We have count samples, and
			 * 2PI radians in a full sine wave. Thus we take
			 * 		(2PI * i / count) * 2048.
			 */
			for (i = 0;i < count; i++)
				mem[i] = sin(2 * M_PI * ((double) i) / count) * BRISTOL_VPO;
			return;
		case 1:
		default:
			/* 
			 * This is a square wave.
			 */
			if (blo.flags & BRISTOL_BLO)
			{
				for (i = 0;i < count; i++)
					mem[i] = blosquare[i];
				return;
			}
			for (i = 0;i < count / 2; i++)
				mem[i] = BRISTOL_VPO * 2 / 3;
			for (;i < count; i++)
				mem[i] = -BRISTOL_VPO * 2 / 3;
			mem[0] = mem[count / 2] = 0;
			mem[count - 1] = mem[1]
				= mem[count / 2 - 1] = mem[count / 2 + 1]
				= mem[2] / 2;
			return;
		case 2:
			/* 
			 * This is a pulse wave (we do not have PWM yet). We should go and
			 * build this out of a subtracted offset ramp wave if BLO is 
			 * enabled.
			 */
			if (blo.flags & BRISTOL_BLO)
			{
				for (i = 0;i < count; i++)
					mem[i] = blopulse[i];
				return;
			}
			for (i = 0;i < count / 5; i++)
				mem[i] = BRISTOL_VPO * 2 / 3;
			for (;i < count; i++)
				mem[i] = -BRISTOL_VPO * 2 / 3;
			mem[0] = mem[count / 2] = 0;
			mem[count - 1] = mem[1]
				= mem[count / 2 - 1] = mem[count / 2 + 1]
				= mem[2] / 2;
			return;
		case 3:
			/* 
			 * This is a ramp wave. We scale the index from -.5 to .5, and
			 * multiply by the range. We go from rear to front to table to make
			 * the ramp wave have a positive leading edge.
			 */
			if (blo.flags & BRISTOL_BLO)
			{
				for (i = 0;i < count; i++)
					mem[i] = bloramp[i];
				return;
			}
			for (i = count - 1;i >= 0; i--)
				mem[i] = (((float) i / count) - 0.5) * BRISTOL_VPO * 2.0;
			mem[0] = 0;
			mem[count - 1] = mem[1]
				= BRISTOL_VPO;
			return;
		case 4:
			/*
			 * Triangular wave. From MIN point, ramp up at twice the rate of
			 * the ramp wave, then ramp down at same rate.
			 */
			if (blo.flags & BRISTOL_BLO)
			{
				for (i = 0;i < count; i++)
					mem[i] = blotriangle[i];
				return;
			}
			for (i = 0;i < count / 2; i++)
				mem[i] = -BRISTOL_VPO
					+ ((float) i * 2 / (count / 2)) * BRISTOL_VPO; 
			for (;i < count; i++)
				mem[i] = BRISTOL_VPO -
					(((float) (i - count / 2) * 2) / (count / 2)) * BRISTOL_VPO;
			return;
		case 5:
			/*
			 * Would like to put in a jagged edged ramp wave. Should make some
			 * interesting strings sounds. If we have BLO enabled then we are
			 * just going to layer the wave with the second harmonic of itself.
			 */
			if (blo.flags & BRISTOL_BLO)
			{
				for (i = 0; i < count; i++)
					mem[i] = (blosaw[i] + blosaw[(i * 2) & 1023]) * 0.5;

				return;
			}
			{
				float accum = -BRISTOL_VPO;

				for (i = 0; i < count / 2; i++)
				{
					mem[i] = accum +
						(0.5 - (((float) i) / count)) * BRISTOL_VPO * 4;
					if (i == count / 8)
						accum = -BRISTOL_VPO * 3 / 4;
					if (i == count / 4)
						accum = -BRISTOL_VPO / 2;
					if (i == count * 3 / 8)
						accum = -BRISTOL_VPO / 4;
				}
				for (; i < count; i++)
					mem[i] = -mem[count - i];
			}
			return;
		case 6:
			/*
			 * Tangiential wave. We limit some of the values, since they do get
			 * excessive. This is only half a tan as well, to maintain the
			 * base frequency.
			 *
			 * This should be dropped.
			 */
			for (i = 0;i < count; i++)
			{
				if ((mem[i] =
					tan(M_PI * ((double) i) / count) * BRISTOL_VPO / 16)
					> BRISTOL_VPO * 8)
					mem[i] = BRISTOL_VPO * 6;
				if (mem[i] < -(BRISTOL_VPO * 6))
					mem[i] = -BRISTOL_VPO * 6;
			}
			return;
		case 7:
			/*
			 * This will be to load a (fixed?) filename. Should probably make
			 * another operator for sample management though, since it would
			 * be better to have pitch shifting incorporated.
			 */
			return;
	}
}

